portal/components/table/_components/virtualRows.tsx

```diff
@@ -0,0 +1,53 @@
+import { flexRender, Row } from '@tanstack/react-table'
+import type { VirtualItem } from '@tanstack/react-virtual'
+
+import { Column } from './column'
+
+type Props<TData> = {
+  loading: boolean
+  rows: Row<TData>[]
+  virtualItems: VirtualItem[]
+}
+
+export function VirtualRows<TData>({
+  loading,
+  rows,
+  virtualItems,
+}: Props<TData>) {
+  if (loading && rows.length === 0) {
+    return null
+  }
+
+  return (
+    <>
+      {virtualItems.map(function (virtualRow) {
+        const row = rows[virtualRow.index] as Row<TData>
```

Comment:
I think this casting can be removed - the compiler should pick it up automatically 🤔 

```suggestion
        const row = rows[virtualRow.index]
```